internal/powerplatform/services/data_record/resource_data_record.go

```diff
@@ -0,0 +1,338 @@
+		MarkdownDescription: "The Power Platform Data Record Resource allows the management of configuration records that are stored in Dataverse as records. This resource is not recommended for managing business data or other data that may be changed by Dataverse users in the context of normal business activities.",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				MarkdownDescription: "Unique id (guid)",
+				Description:         "Unique id (guid)",
+				Computed:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"environment_id": schema.StringAttribute{
+				Description: "Id of the Dynamics 365 environment",
+				Required:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"table_logical_name": schema.StringAttribute{
+				Description: "Logical name of the data record table",
+				Required:    true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"columns": schema.DynamicAttribute{
+				Description: "Columns of the data record table",
+				Required:    true,
+			},
+		},
+	}
+}
+
+func (r *DataRecordResource) Configure(ctx context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	clientApi := req.ProviderData.(*api.ProviderClient).Api
+	if clientApi == nil {
+		resp.Diagnostics.AddError(
+			"Unexpected Resource Configure Type",
+			fmt.Sprintf("Expected *http.Client, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+
+		return
+	}
+	r.DataRecordClient = NewDataRecordClient(clientApi)
+}
+
+func (r *DataRecordResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var state *DataRecordResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &state)...)
+
+	var plan DataRecordResourceModel
+	resp.State.Get(ctx, &plan)
+
+	tflog.Debug(ctx, fmt.Sprintf("CREATE RESOURCE START: %s", r.ProviderTypeName))
+
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	plan.Id = types.StringValue(plan.Id.ValueString())
+	plan.EnvironmentId = types.StringValue(plan.EnvironmentId.ValueString())
+	plan.TableLogicalName = types.StringValue(plan.TableLogicalName.ValueString())
+	plan.Columns = types.DynamicValue(plan.Columns)
+
+	var mapColumns map[string]interface{}
+
+	jsonColumns, _ := json.Marshal(plan.Columns.String())
+	unquotedJsonColumns, _ := strconv.Unquote(string(jsonColumns))
+	json.Unmarshal([]byte(unquotedJsonColumns), &mapColumns)
+
+	dr, err := r.DataRecordClient.ApplyDataRecord(ctx, plan.Id.ValueString(), plan.EnvironmentId.ValueString(), plan.TableLogicalName.ValueString(), mapColumns)
```

Comment:
after applydatarecord you should call convertColumnsToState to save what you've got back to the state, because backend could potentialy return you something that looks different than what tf file has like address_composite1 field